### ISSUE TITLE
http.get fails on Android and FreeBSD

### ISSUE THREAD

**Issue by javihernandez** [joyent/node#8540]
Hi,

when making a simple http request I'm getting the following error:

<pre>
Error: getaddrinfo EAI_BADFLAGS
    at Object.exports._errnoException (util.js:742:11)
    at errnoException (dns.js:46:15)
    at Object.onlookup [as oncomplete] (dns.js:91:26)
</pre>

Did a few tests and tried to follow the error through the code, but at this point all I can say is that I'm lost. Any ideas? As usually, I'll be happy to help on fixing this.

Resulting object from calling http.get can be found [here](http://pastebin.com/VAYJXPaE)

**Comment by josser** [comment]
Very similar issue on freebsd 10: 
```
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: getaddrinfo EAI_BADFLAGS
    at Object.exports._errnoException (util.js:746:11)
    at errnoException (dns.js:49:15)
    at GetAddrInfoReqWrap.onlookup [as oncomplete] (dns.js:94:26)
```

Code: 

 ```javascript 
http = require('http');
http.get("www.google.com", function(a,b,c) {}); 
```
man 3 getaddrinfo https://gist.github.com/josser/ccad69a7c767aa0d3159 

node --version
v0.12.0

**Comment by josser** [comment]
Also, I don't have such error on OS X, but it have a bit more extended list of AI_FLAGS: 
Additional flags which I found in os x but didn't find in freebsd: 

AI_ALL          If the AI_ALL bit is set with the AI_V4MAPPED bit, then getaddrinfo() shall return all matching IPv6 and IPv4 addresses.  The AI_ALL bit without the AI_V4MAPPED bit is ignored.

AI_V4MAPPED     If the AI_V4MAPPED flag is specified along with an ai_family of AF_INET6, then getaddrinfo() shall return IPv4-mapped IPv6 addresses on finding no matching IPv6 addresses ( ai_addrlen shall be 16).  The AI_V4MAPPED flag shall be ignored unless ai_family equals AF_INET6.

**Comment by leonkunert** [comment]
I have a similar error after updating from node-0.10.35 to 0.12.0 on Freebsd 10.1.

```
{ [Error: getaddrinfo EAI_BADFLAGS]
  code: 'EAI_BADFLAGS',
  errno: 'EAI_BADFLAGS',
  syscall: 'getaddrinfo',
  hostname: 'xxx' }
```

**Comment by pct** [comment]
Please see the patch, Thanks.

https://github.com/iojs/io.js/commit/04bea9f9c224d9f863f671a1ad52a3f392b292cb

https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=197539

**Comment by misterdjules** [comment]
More background for the issue on FreeBSD [here](http://lists.freebsd.org/pipermail/freebsd-bugs/2008-February/028260.html), with the related source change [here](https://svnweb.freebsd.org/base/head/lib/libc/net/getaddrinfo.c?r1=172052&r2=175955).

https://github.com/iojs/io.js/commit/04bea9f9c224d9f863f671a1ad52a3f392b292cb seems it would definitely fix this issue, however I'm wondering whether we should fix this in `dns.lookup` or in `net.connect`.

In my opinion the question boils down to: do we want explicit calls to `dns.lookup` with `V4MAPPED` set to success silently on FreeBSD? Users of FreeBSD might be confused by the fact that `dns.lookup` would not return IPv4-mapped addresses when they explicitly asked for it.

Instead, we could remove the `V4MAPPED` flag in `net.connect` which is the only place in node.js' core where it's set. That would fix this issue, and explicit uses of `dns.lookup` with `V4MAPPED` would fail on FreeBSD, making it clear that FreeBSD users should not expect IPv4-mapped to work.

@joyent/node-coreteam Thoughts?

**Comment by piscisaureus** [comment]
> Instead, we could remove the V4MAPPED flag in net.connect which is the only place in node.js' core where it's set. That would fix this issue, and explicit uses of dns.lookup with V4MAPPED would fail on FreeBSD, making it clear that FreeBSD users should not expect IPv4-mapped to work.

That's a possibility, but also an API change. I guess for node that means this goes into v0.14.0

**Comment by misterdjules** [comment]
@piscisaureus I don't understand why it would be an API change, could you please elaborate?

The other candidate for a fix I had in mind is available here: https://gist.github.com/misterdjules/8b451749630de0b17218. In my opinion, this is a behavior change that only affects a single platform where the previous behavior is not available and fails all the time, so I would qualify it as a bug fix, but I may be missing something.

Also, again to make sure we're on the same page, if https://gist.github.com/misterdjules/8b451749630de0b17218 qualifies as a breaking API change, then my understanding is that https://github.com/iojs/io.js/commit/04bea9f9c224d9f863f671a1ad52a3f392b292cb would also be considered as a breaking change.

Thanks!

**Comment by piscisaureus** [comment]
@misterdjules 
I see what you mean. I would say your fix is better. @bnoordhuis @indutny wdyt?

**Comment by bnoordhuis** [comment]
I'm working from memory here but Fedor's patch disables AI_V4MAPPED iff it's not an AAAA query and the platform is FreeBSD.  Julien's patch looks wrong to me because it disables it unconditionally on FreeBSD, regardless of the query type.

Side note: OpenBSD doesn't support AI_V4MAPPED at all, see io.js/iojs@ca039b46.

**Comment by misterdjules** [comment]
@bnoordhuis Ok, based on this post: http://lists.freebsd.org/pipermail/freebsd-bugs/2008-February/028260.html, I thought FreeBSD also did not support `AI_V4MAPPED` at all (regardless of the query type) in recent versions.

It seems that the code example in this message does an AAAA query with `AI_V4MAPPED` and `getaddrinfo` returns `EAI_BADFLAGS`. However, I don't have a FreeBSD setup I can test this on, so I'm only guessing.

**Comment by bnoordhuis** [comment]
That mailing list post is about FreeBSD 7, which is ancient.  io.js's minimum requirement is FreeBSD 10 and that seems to support AI_V4MAPPED fine for AAAA queries.

**Comment by indutny** [comment]
I think we better remove this flag at the query time since V4MAPPED is
exposed to users and might cause platform imcompatibility in user code

**Comment by misterdjules** [comment]
@bnoordhuis I spun up a FreeBSD 10.1 machine it seems that `AI_V4MAPPED` is not supported even for AAAA queries:
```
$ uname -a
FreeBSD aa021983-1c97-6978-a0f3-a6f715347d9c 10.1-RELEASE FreeBSD 10.1-RELEASE #0 r274401: Tue Nov 11 21:02:49 UTC 2014     root@releng1.nyi.freebsd.org:/usr/obj/usr/src/sys/GENERIC  amd64
$ cat test-mapped.c 
#include <sys/socket.h>
#include <sys/types.h>

#include <netdb.h>
#include <netinet/in.h>
#include <stdio.h>
#include <string.h>


int main(int argc, char** argv)
{
	struct addrinfo hints;
	struct addrinfo* res;
	int rv;

	memset(&hints,0,sizeof(hints));
	hints.ai_flags = AI_V4MAPPED; 
	hints.ai_family = AF_INET6;
	hints.ai_socktype = SOCK_STREAM;

	if( ( rv = getaddrinfo("www.google.com","http",&hints,&res) ) ) {
		printf("Error: %s.\n",gai_strerror(rv));
		return 1;
	}

	printf("Got address(es)\n");
	freeaddrinfo(res);
	return 0;
}
$ cc test-mapped.c -o test-mapped
$ ./test-mapped 
Error: Invalid value for ai_flags.
$
```
Am I missing something?

**Comment by bnoordhuis** [comment]
Is IPv6 enabled on the machine?  FreeBSD has some interesting failure modes in that respect.

If it's not that, I defer to @indutny.  He's the author of the patch and the main tester / maintainer of the FreeBSD build.

**Comment by indutny** [comment]
Ben, I do believe that I did that query check because you was insisting ;)
didn't seen any way this flag could be accosted when I was reading the
kernel code.

**Comment by misterdjules** [comment]
@bnoordhuis IPv6 seems to be enabled on this machine:
```
$ ifconfig | grep inet6
	inet6 ::1 prefixlen 128 
	inet6 fe80::1%lo0 prefixlen 64 scopeid 0x3 
$
```

**Comment by misterdjules** [comment]
I had some more time to continue my investigations. So it seems that FreeBSD 10.1 does not support `AI_V4MAPPED` at all, even for AAAA queries. Here's a slightly modified version of the small C program I mentioned in the previous comments:
```
$ cat test.c
#include <sys/socket.h>
#include <sys/types.h>

#include <netdb.h>
#include <netinet/in.h>
#include <stdio.h>
#include <string.h>
#include <arpa/inet.h>

int main(int argc, char** argv)
{
    struct addrinfo hints;
    struct addrinfo* res = NULL, *cur_res = NULL;
    int r;
    char str_addr[INET6_ADDRSTRLEN];

    memset(&hints,0,sizeof(hints));
    hints.ai_flags = AI_V4MAPPED; 
    hints.ai_family = AF_INET6;
    hints.ai_socktype = SOCK_STREAM;

    printf("Looking up www.google.com...\n");

    if ((r = getaddrinfo("www.google.com","http", &hints, &res))) {
        printf("Error: %s.\n",gai_strerror(r));
        return 1;
    } else {
      printf("Got address(es)\n");
      for (cur_res = res; cur_res; cur_res = res->ai_next) {
        struct sockaddr_in6 *addr;
        addr = (struct sockaddr_in6*)cur_res->ai_addr; 
        inet_ntop(AF_INET6, &addr->sin6_addr, str_addr,
                  sizeof(str_addr));
        printf("IPv6 address: %s\n", str_addr);
      }
    }
    
    freeaddrinfo(res);
    return 0;
}
$ cc test.c -o test
$ ./test 
Looking up www.google.com...
Error: Invalid value for ai_flags.
$ diff -uN test.c test-no-v4mapped.c 
--- test.c	2015-04-17 19:21:57.000000000 +0000
+++ test-no-v4mapped.c	2015-04-17 20:34:53.000000000 +0000
@@ -15,7 +15,7 @@
     char str_addr[INET6_ADDRSTRLEN];
 
     memset(&hints,0,sizeof(hints));
-    hints.ai_flags = AI_V4MAPPED; 
+    //hints.ai_flags = AI_V4MAPPED; 
     hints.ai_family = AF_INET6;
     hints.ai_socktype = SOCK_STREAM;
 
$ cc test-no-v4mapped.c -o test-no-v4mapped
$ ./test-no-v4-mapped
Looking up www.google.com...
Got address(es)
IPv6 address: 2607:f8b0:4002:c01::69
$
```
which is also reflected in the behavior of latest io.js when using `dns.lookup` with `family: 6` and `hints: dns.V4MAPPED`:
```
root@381f0506-b6c5-c14a-db53-b02109976e52:/tmp/io.js # cat dns-ipv6-mapped.js 
var dns = require('dns');
dns.lookup('www.google.com', { family: 6, hints: dns.V4MAPPED }, function onLookup(err) {
  console.log('err:', err);
});
$ ./iojs --version
v1.7.2
$ ./iojs dns-ipv6-mapped.js 
err: { [Error: getaddrinfo EAI_BADFLAGS]
  code: 'EAI_BADFLAGS',
  errno: 'EAI_BADFLAGS',
  syscall: 'getaddrinfo',
  hostname: 'www.google.com' }
$
```
So as I understand it, https://github.com/iojs/io.js/commit/04bea9f9c224d9f863f671a1ad52a3f392b292cb does not fix `dns.lookup` entirely regarding the usage of `dns.V4MAPPED` on FreeBSD. In its current status, it seems to only fix the `net.connect` use case, but by changing code that is not necessarily directly related. If I understand the original intention correctly, this change would need to be modified to also clear `dns.V4MAPPED` when `family === 6`.

So it seems to me we have several different ways to fix this problem:

1. Not using `dns.V4MAPPED` in `net.connect` _only on FreeBSD_ and not changing `dns.lookup`. __Pros:__ This fixes the `net.connect` use case. Users using `dns.V4MAPPED`  in `dns.lookup` will get an error on FreeBSD clearly indicating that it's not supported instead of having to find out why it doesn't return mapped addresses. __Cons:__ code that uses `dns.lookup` and the `dns.V4MAPPED` flag running on several different platforms including FreeBSD might need to be modified to special-case FreeBSD. 

2. Using https://github.com/iojs/io.js/commit/04bea9f9c224d9f863f671a1ad52a3f392b292cb as is, with the problems highlighted above.

3. Using https://github.com/iojs/io.js/commit/04bea9f9c224d9f863f671a1ad52a3f392b292cb and clearing the `dns.V4MAPPED` flag regardless of the family on FreeBSD. __Pros:__ that fixes all errors that FreeBSD users will get when using `dns.V4MAPPED`. __Cons:__ That can leave FreeBSD users wondering why mapped addresses are not returned when they explicitly asked for it.

4. Using a fix similar to [the one for OpenBSD](Using https://github.com/iojs/io.js/commit/04bea9f9c224d9f863f671a1ad52a3f392b292cb), as it seems that `AI_V4MAPPED` is not supported at all.

@joyent/node-coreteam We need to make a decision and I want your input on this. My preference goes to 1 because it fixes the most common use case and it doesn't fail silently for the other explicit `dns.lookup` use cases, giving the right information to the users so that they know what's happening.

**Comment by trevnorris** [comment]
1 sounds good

**Comment by misterdjules** [comment]
@trevnorris Thank you for the feedback, created #18204.

**Comment by misterdjules** [comment]
Downgrading to P-2 since it seems to be fixed in FreeBSD ports.

**Comment by misterdjules** [comment]
Fixed by 0e392f3, thank you everyone!
